File: sdk/python/examples/generate_fixture.py
"""Generate the SDK-emitted fixture the platform's cross-stack test consumes.

Writes the research-pipeline scenario (see scenario.py) to
``frontend/lib/__tests__/fixtures/sdk-emitted-events.jsonl`` via the JSONL transport — the
same append-only format an EventSource reads. The platform test
``frontend/lib/__tests__/sdk-events.test.ts`` then proves identical analytics.

    python examples/generate_fixture.py [output_path]
"""

from __future__ import annotations

import os
import sys

# Make the SDK importable when run directly from sdk/python/.
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from observation_sdk import JsonlTransport, ObservationClient  # noqa: E402
from examples.scenario import emit_scenario  # noqa: E402

_DEFAULT_OUT = os.path.normpath(
    os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "..", "..", "..",
        "frontend", "lib", "__tests__", "fixtures", "sdk-emitted-events.jsonl",
    )
)


def write_fixture(out_path: str) -> int:
    """Write the scenario to ``out_path`` (mode='w', deterministic). Returns event count."""
    client = ObservationClient(
        JsonlTransport(out_path, mode="w"),
        application_name="research-pipeline-demo",
        environment="demo",
    )
    try:
        events = emit_scenario(client)
    finally:
        client.close()
    return len(events)


def main() -> None:
    out_path = sys.argv[1] if len(sys.argv) > 1 else _DEFAULT_OUT
    n = write_fixture(out_path)
    print(f"Wrote {n} SDK-emitted ObservationEvents to {out_path}")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/generateFixture.ts
/**
 * Generate the TypeScript-SDK-emitted fixture the platform's cross-stack parity test consumes.
 *
 * Writes the research-pipeline scenario (see scenario.ts) to
 * `frontend/lib/__tests__/fixtures/sdk-emitted-events.typescript.jsonl` via the JSONL transport
 * — the same append-only format an EventSource reads. The platform test
 * `frontend/lib/__tests__/sdk-parity.test.ts` then proves identical analytics AND field-for-field
 * parity with the Python-emitted fixture (only `metadata.sdk` differs).
 *
 *     npm run gen:fixture            # build, then write to the default path
 *     node dist/examples/generateFixture.js [outputPath]
 */

import { existsSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { ObservationClient } from "../src/client.js";
import { JsonlTransport } from "../src/transport.js";
import { emitScenario } from "./scenario.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));

/**
 * Find the repo root by ascending until a directory holds both `sdk/` and `frontend/`.
 * Robust whether this runs from source (`examples/`) or compiled (`dist/examples/`).
 */
function repoRoot(start: string): string {
  let dir = start;
  for (;;) {
    if (existsSync(path.join(dir, "sdk")) && existsSync(path.join(dir, "frontend"))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) throw new Error("could not locate repo root (no sdk/ + frontend/ above " + start + ")");
    dir = parent;
  }
}

const DEFAULT_OUT = path.join(
  repoRoot(HERE),
  "frontend", "lib", "__tests__", "fixtures", "sdk-emitted-events.typescript.jsonl",
);

/** Write the scenario to `outPath` (mode 'w', deterministic). Returns the event count. */
export function writeFixture(outPath: string): number {
  const client = new ObservationClient(new JsonlTransport(outPath, { mode: "w" }), {
    applicationName: "research-pipeline-demo",
    environment: "demo",
  });
  try {
    return emitScenario(client).length;
  } finally {
    void client.close();
  }
}

function main(): void {
  const outPath = process.argv[2] ?? DEFAULT_OUT;
  const n = writeFixture(outPath);
  console.log(`Wrote ${n} SDK-emitted ObservationEvents to ${outPath}`);
}

// Run main() only when invoked directly (not when imported by a test).
const invokedDirectly = process.argv[1] !== undefined && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (invokedDirectly) {
  main();
}
